streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBVersionedStoreSegmentValueFormatterTest.java

```diff
@@ -0,0 +1,316 @@
+        final SegmentValue deserialized = RocksDBVersionedStoreSegmentValueFormatter.deserialize(serialized);
+
+        verifySegmentContents(deserialized, testCase);
+    }
+
+    @Test
+    public void shouldBuildWithInsertLatest() {
+        final SegmentValue segmentValue = buildSegmentWithInsertLatest(testCase);
+
+        verifySegmentContents(segmentValue, testCase);
+    }
+
+    @Test
+    public void shouldBuildWithInsertEarliest() {
+        final SegmentValue segmentValue = buildSegmentWithInsertEarliest(testCase);
+
+        verifySegmentContents(segmentValue, testCase);
+    }
+
+    @Test
+    public void shouldInsertAtIndex() {
+        if (testCase.isDegenerate) {
+            // cannot insert into degenerate segment
+            return;
+        }
+
+        // test inserting at each possible index
+        for (int insertIdx = 0; insertIdx <= testCase.records.size(); insertIdx++) {
+            // build record to insert
+            final long newRecordTimestamp;
+            if (insertIdx == 0) {
+                newRecordTimestamp = testCase.records.get(0).timestamp + 1;
+                if (newRecordTimestamp == testCase.nextTimestamp) {
+                    // cannot insert because no timestamp exists between last record and nextTimestamp
+                    continue;
+                }
+            } else {
+                newRecordTimestamp = testCase.records.get(insertIdx - 1).timestamp - 1;
+                if (newRecordTimestamp < 0 || (insertIdx < testCase.records.size() && newRecordTimestamp == testCase.records.get(insertIdx).timestamp)) {
+                    // cannot insert because timestamps of existing records are adjacent
+                    continue;
+                }
+            }
+            final TestRecord newRecord = new TestRecord("new".getBytes(), newRecordTimestamp);
+
+            final SegmentValue segmentValue = buildSegmentWithInsertLatest(testCase);
+
+            // insert() first requires a call to find()
+            if (insertIdx > 0) {
+                segmentValue.find(testCase.records.get(insertIdx - 1).timestamp, false);
+            }
+            segmentValue.insert(newRecord.timestamp, newRecord.value, insertIdx);
+
+            // create expected results
+            final List<TestRecord> expectedRecords = new ArrayList<>(testCase.records);
+            expectedRecords.add(insertIdx, newRecord);
+
+            verifySegmentContents(segmentValue, new TestCase("expected", testCase.nextTimestamp, expectedRecords));
+        }
+    }
+
+    @Test
+    public void shouldUpdateAtIndex() {
+        if (testCase.isDegenerate) {
+            // cannot update degenerate segment
+            return;
+        }
+
+        // test updating at each possible index
+        for (int updateIdx = 0; updateIdx < testCase.records.size(); updateIdx++) {
+            // build updated record
+            long updatedRecordTimestamp = testCase.records.get(updateIdx).timestamp - 1;
```

Comment:
If I understand the test correctly, we try to update the timestamp and fall back to use the same timestamp if there is no "gap". It seems we don't test both case where the timestamp is increase or decreased but use decreased, and only "fall back" to increase does not work. Should we try to test both cases instead?

Reply:
There are enough test cases with adjacent timestamps that both cases are exercised. I don't think we need to explicitly test both in all positions, since the underlying implementation doesn't care what the new timestamps are. This logic here is just to ensure that the timestamps we end up with are valid, i.e., in strictly increasing order.